File: Components/search_mu_polynomial.m
function [output1,output2] = search_mu_polynomial(varargin)
% The polynomial mutation.

%------------------------------Reference-----------------------------------
% Deb K, Goyal M. A combined genetic adaptive search (GeneAS) for 
% engineering design[J]. Computer Science and Informatics, 1996, 26: 30-45.
%------------------------------Copyright-----------------------------------
% Copyright (C) <2023>  <Swarm Intelligence Lab>

% AutoOptLib is a free software. You can use, redistribute, and/or modify
% it under the terms of the GNU General Public License as published by the 
% Free Software Foundation, either version 3 of the License, or any later 
% version. 

% Please reference the paper below if using AutoOptLib in your publication:
% @article{zhao2023autooptlib,
%  title={AutoOptLib: A Library of Automatically Designing Metaheuristic 
%         Optimization Algorithms in Matlab},
%  author={Zhao, Qi and Yan, Bai and Hu, Taiwei and Chen, Xianglong and 
%          Yang, Jian and Shi, Yuhui},
%  journal={arXiv preprint 	arXiv:2303.06536},
%  year={2023}
% }
%--------------------------------------------------------------------------

mode = varargin{end};
switch mode
    case 'execute'
        Parent  = varargin{1};
        Problem = varargin{2};
        Para    = varargin{3};
        Aux     = varargin{4};
        ProbM   = Para(1);
        DisM    = Para(2);
        
        if ~isnumeric(Parent)
            Offspring = Parent.decs;           
        else
            Offspring = Parent;
        end    
        [N,D] = size(Offspring);

        Lower = repmat(Problem.bound(1,:),N,1);
        Upper = repmat(Problem.bound(2,:),N,1);

        Site  = rand(N,D) < ProbM;
        mu    = rand(N,D);
        temp  = Site & mu<=0.5;
        Offspring(temp) = Offspring(temp)+(Upper(temp)-Lower(temp)).*((2.*mu(temp)+(1-2.*mu(temp)).*...
            (1-(Offspring(temp)-Lower(temp))./(Upper(temp)-Lower(temp))).^(DisM+1)).^(1/(DisM+1))-1);
        temp = Site & mu>0.5;
        Offspring(temp) = Offspring(temp)+(Upper(temp)-Lower(temp)).*(1-(2.*(1-mu(temp))+2.*(mu(temp)-0.5).*...
            (1-(Upper(temp)-Offspring(temp))./(Upper(temp)-Lower(temp))).^(DisM+1)).^(1/(DisM+1)));
        output1 = Offspring;
        output2 = Aux;
    
    case 'parameter'
        output1 = [0,0.3;20,40]; % mutation probability and distribution

    case 'behavior'
        output1 = {'LS','small','large';'GS','large','small'}; % small probabilities and large distributions perform local search

    case 'algorithm'
        Parent    = varargin{1};
        bound     = varargin{2};
        ProbM     = 1;
        DisM      = 30;
        Offspring = Parent;
        [N,D] = size(Offspring);

        Lower = repmat(bound(1,:),N,1);
        Upper = repmat(bound(2,:),N,1);
        Site  = rand(N,D) < ProbM;
        mu    = rand(N,D);
        temp  = Site & mu<=0.5;
        Offspring(temp) = Offspring(temp)+(Upper(temp)-Lower(temp)).*((2.*mu(temp)+(1-2.*mu(temp)).*...
            (1-(Offspring(temp)-Lower(temp))./(Upper(temp)-Lower(temp))).^(DisM+1)).^(1/(DisM+1))-1);
        temp = Site & mu>0.5;
        Offspring(temp) = Offspring(temp)+(Upper(temp)-Lower(temp)).*(1-(2.*(1-mu(temp))+2.*(mu(temp)-0.5).*...
            (1-(Upper(temp)-Offspring(temp))./(Upper(temp)-Lower(temp))).^(DisM+1)).^(1/(DisM+1)));
        output1 = Offspring;
end

if ~exist('output1','var')
    output1 = [];
end
if ~exist('output2','var')
    output2 = [];
end
end

File: AutoOptLib for Python/components/search_mu_polynomial.py
"""Python translation of search_mu_polynomial."""

from __future__ import annotations

import numpy as np


def search_mu_polynomial(*args):
    mode = args[-1]
    if mode == "execute":
        parent = args[0]
        problem = args[1] if len(args) > 1 else None
        para = args[2] if len(args) > 2 else None
        aux = args[3] if len(args) > 3 else None

        if not isinstance(parent, (np.ndarray, list, tuple)):
            decs = getattr(parent, "decs", None)
            offspring = decs() if callable(decs) else (decs if decs is not None else parent)
        else:
            offspring = parent
        offspring = np.asarray(offspring, dtype=float)
        n, d = offspring.shape

        prob_m = 0.1
        dis_m = 30.0
        if para is not None:
            arr = np.asarray(para).reshape(-1)
            if arr.size > 0:
                prob_m = float(arr[0])
            if arr.size > 1:
                dis_m = float(arr[1])

        bound = getattr(problem, "bound")
        lower = np.asarray(bound[0], dtype=float)
        upper = np.asarray(bound[1], dtype=float)
        lower = np.repeat(lower.reshape(1, -1), n, axis=0)
        upper = np.repeat(upper.reshape(1, -1), n, axis=0)

        site = np.random.rand(n, d) < prob_m
        mu = np.random.rand(n, d)
        temp = site & (mu <= 0.5)
        if np.any(temp):
            offspring[temp] = offspring[temp] + (upper[temp] - lower[temp]) * (
                (2.0 * mu[temp] + (1.0 - 2.0 * mu[temp]) * (1.0 - (offspring[temp] - lower[temp]) / (upper[temp] - lower[temp])) ** (dis_m + 1.0)) ** (1.0 / (dis_m + 1.0)) - 1.0
            )
        temp = site & (mu > 0.5)
        if np.any(temp):
            offspring[temp] = offspring[temp] + (upper[temp] - lower[temp]) * (
                1.0 - (2.0 * (1.0 - mu[temp]) + 2.0 * (mu[temp] - 0.5) * (1.0 - (upper[temp] - offspring[temp]) / (upper[temp] - lower[temp])) ** (dis_m + 1.0)) ** (1.0 / (dis_m + 1.0))
            )
        return offspring, aux

    if mode == "parameter":
        return [[0, 0.3], [20, 40]], None

    if mode == "behavior":
        return [["LS", "small", "large"], ["GS", "large", "small"]], None

    if mode == "algorithm":
        parent = args[0]
        bound = np.asarray(args[1], dtype=float)
        prob_m = 1.0
        dis_m = 30.0
        offspring = np.asarray(parent, dtype=float)
        n, d = offspring.shape
        lower = np.repeat(bound[0].reshape(1, -1), n, axis=0)
        upper = np.repeat(bound[1].reshape(1, -1), n, axis=0)
        site = np.random.rand(n, d) < prob_m
        mu = np.random.rand(n, d)
        temp = site & (mu <= 0.5)
        if np.any(temp):
            offspring[temp] = offspring[temp] + (upper[temp] - lower[temp]) * (
                (2.0 * mu[temp] + (1.0 - 2.0 * mu[temp]) * (1.0 - (offspring[temp] - lower[temp]) / (upper[temp] - lower[temp])) ** (dis_m + 1.0)) ** (1.0 / (dis_m + 1.0)) - 1.0
            )
        temp = site & (mu > 0.5)
        if np.any(temp):
            offspring[temp] = offspring[temp] + (upper[temp] - lower[temp]) * (
                1.0 - (2.0 * (1.0 - mu[temp]) + 2.0 * (mu[temp] - 0.5) * (1.0 - (upper[temp] - offspring[temp]) / (upper[temp] - lower[temp])) ** (dis_m + 1.0)) ** (1.0 / (dis_m + 1.0))
            )
        return offspring, None

    raise ValueError(f"Unsupported mode: {mode}")
